erigon-lib/downloader/downloader.go

```diff
@@ -117,23 +138,67 @@ func New(ctx context.Context, cfg *downloadercfg.Cfg, dirs datadir.Dirs, logger
 	}
 
 	d := &Downloader{
-		cfg:               cfg,
-		db:                db,
-		pieceCompletionDB: c,
-		folder:            m,
-		torrentClient:     torrentClient,
-		statsLock:         &sync.RWMutex{},
-		webseeds:          &WebSeeds{logger: logger, verbosity: verbosity, downloadTorrentFile: cfg.DownloadTorrentFilesFromWebseed, torrentsWhitelist: lock.Downloads},
-		logger:            logger,
-		verbosity:         verbosity,
-		torrentFiles:      &TorrentFiles{dir: cfg.Dirs.Snap},
-		snapshotLock:      lock,
+		cfg:                 cfg,
+		db:                  db,
+		pieceCompletionDB:   c,
+		folder:              m,
+		torrentClient:       torrentClient,
+		lock:                &sync.RWMutex{},
+		webseeds:            &WebSeeds{logger: logger, verbosity: verbosity, downloadTorrentFile: cfg.DownloadTorrentFilesFromWebseed, torrentsWhitelist: lock.Downloads},
+		logger:              logger,
+		verbosity:           verbosity,
+		torrentFiles:        &TorrentFiles{dir: cfg.Dirs.Snap},
+		snapshotLock:        lock,
+		webDownloadInfo:     map[string]webDownloadInfo{},
+		webDownloadSessions: map[string]*RCloneSession{},
+		downloading:         map[string]struct{}{},
+	}
+
+	if cfg.ClientConfig.DownloadRateLimiter != nil {
+		downloadLimit := cfg.ClientConfig.DownloadRateLimiter.Limit()
+		d.downloadLimit = &downloadLimit
 	}
 
 	d.webseeds.torrentFiles = d.torrentFiles
 	d.ctx, d.stopMainLoop = context.WithCancel(ctx)
 
 	if cfg.AddTorrentsFromDisk {
+		var downloadMismatches []string
+
+		for _, download := range lock.Downloads {
```

Comment:
if you add from disk only files which are in lock.json file - then i don't understand meaning of lock.json file. 
If we add from disk only files which are in lock - then we will not add files which erigon created by himself (not downloaded)? I my head lock.json storing only downloaded files, no? (need somewhere in code write meaning of lock.json - what is there and what is not there).

Reply:
ok - I have started to add some description in the readme.md.  Becuase I think the downloader is now complex enough to need an explanation.

What I think should be happening at the moment is that the code does the following:

1. For the snapshot lock files - it ignores the .seg.torrent's because these are of unknown provenance.  (I think they as redundant, but have not removed them entirely as I don't know what they are used for.
2. Fur locally produced files the process should work as it currently does (if it doesn't this is a bug) - I did test this was working a week ago - but with recent changes maybe it has regressed.  I'll retest it.

I think that in the future we should add an `uploads` section to the lock.  Then we can use that for managing upload state.  The current uploader code does this by reading the manifest of the remote disk - but it would be good to do better state manager here I think.  But this is down the road.

Reply:
What this code is doing it reconciling the state between the set of downloaded files: the contents of either snapshot-json.lock or config.toml and the data downloaded and stored in the downloader db.  It making sure that the preconfigured files actually go downloaded.

Any file not in this list (by file name) is ignored by this code.  It means that the local node can't override pre-configured files - it can only add files.  I think this should always be true for normal operations and it seems like a base level integrity check than needs to be performed.

The way the code operates, if you want to make changes to this for development purposes, or because there is some kind of discrepancy you need to override the contents of the lock file.  

The purpose of the logging is to give you the hashes.  I'm not sure if you are asking me to make code changes here ?

Reply:
I didn't read your comment yet. Will reply later. Here is small context. 

About `.torrent`:
- `.torrent` contains: 1 `sha1` hash for each `2mb` piece of file. BitTorrent network-protocol working on this hashes. to download data. `datadir/downloader` db schema: `piece_hash -> bool_valid_or_not`.
- `.torrent` has total file length
- `.torrent` also may include: list of public trackers where this `.torrent` hosted now (public trackers have ttl), list of webseeds, etc... But we don't add it to files on disk.
- All bittorrent clients/seedboxes/software working with .torrent files - so, it's basic primitive - this is reason why I also creating/reading them. 


About downloader: 
- it has several cases: 1) has .torrent on disk and no files  -> download files. 2) has files and not .torrent -> create .torrent by file. 3) has files and .torrent but `datadir/downloader` folder was manually removed  -> re-gen db without downloading. 
- this cases may appear by many reasons: user restored from backup, user has problem and I told him delete some file, etc...
- later we added `webseeds` (async downloading .torrent files from) and it interfere with all 3 cases
- later we added `lock` file which interfere with all 3 cases
- `downloader --verify` flag is similar to removal of `datadir/downloader`.